linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,455 @@
+        int64_t current_index;
+
+        do
+        {
+            do
+            {
+                if (clock_gettime(CLOCK_REALTIME, &ts) == -1)
+                {
+                    LogError("Failure getting time from clock_gettime");
+                }
+                else
+                {
+                    // Setup the timeout for the semaphore to 2 sec
+                    ts.tv_sec += 2;
+                    if (sem_timedwait(&threadpool->semaphore, &ts) == 0)
+                    {
+                        break;
+                    }
+                }
+            } while (interlocked_add(&threadpool->quitting, 0) != 1);
+
+            if (interlocked_add(&threadpool->quitting, 0) != 1)
+            {
+                int32_t existing_count = interlocked_add(&threadpool->task_array_count, 0);
+
+                // We have a new item
+                current_index = interlocked_add_64(&threadpool->consume_idx, 0) % existing_count;
+
+                // Is this index waiting for us to complete it
+                srw_lock_acquire_shared(threadpool->srw_lock);
+                TASK_RESULT curr_task_result = interlocked_compare_exchange(&threadpool->task_array[current_index].task_state, TASK_WORKING, TASK_WAITING);
+                srw_lock_release_shared(threadpool->srw_lock);
+                if (TASK_WAITING == curr_task_result)
+                {
+                    int64_t consume_pos = interlocked_compare_exchange_64(&threadpool->consume_idx, current_index+1, current_index);
+                    if (consume_pos == current_index)
+                    {
+                        srw_lock_acquire_shared(threadpool->srw_lock);
+                        THREADPOOL_TASK* temp_item = &threadpool->task_array[consume_pos];
+                        srw_lock_release_shared(threadpool->srw_lock);
+
+                        // Get the item and make sure it's valid
+                        if (temp_item != NULL)
+                        {
+                            temp_item->task_func(temp_item->task_param);
+                        }
+                        // I'm done with this item mark it as complete
+                        srw_lock_acquire_shared(threadpool->srw_lock);
+                        interlocked_exchange(&threadpool->task_array[consume_pos].task_state, TASK_NOT_USED);
+                        srw_lock_release_shared(threadpool->srw_lock);
+                    }
+                }
+            }
+        } while (interlocked_add(&threadpool->quitting, 0) != 1);
+        (void)interlocked_decrement(&threadpool->thread_count);
+    }
+}
+
+static int reallocate_threadpool_array(THREADPOOL* threadpool)
+{
+    int result;
+    {
+        // Lock the array here
+        srw_lock_acquire_exclusive(threadpool->srw_lock);
+
+        int32_t existing_count = interlocked_add(&threadpool->task_array_count, 0);
+
+        // Double the items
+        (void)interlocked_exchange(&threadpool->task_array_count, existing_count*2);
```

Comment:
>interlocked_exchange [](http://example.com/codeflow?start=14&length=20)

This has to be interlocked compare exchange (because another thread could come and change it in the meanwhile also.